File: generic-utilities/src/Repository.ts
/**
 * Base entity interface that all entities must implement
 */
export interface Entity {
    id: string | number;
}

/**
 * Repository interface defining CRUD operations
 * @template T - Entity type that extends Entity
 */
export interface Repository<T extends Entity> {
    findById(id: T['id']): Promise<T | null>;
    findAll(): Promise<T[]>;
    findBy<K extends keyof T>(field: K, value: T[K]): Promise<T[]>;
    create(entity: Omit<T, 'id'>): Promise<T>;
    update(id: T['id'], updates: Partial<Omit<T, 'id'>>): Promise<T | null>;
    delete(id: T['id']): Promise<boolean>;
}

/**
 * In-memory implementation of the Repository interface
 * @template T - Entity type that extends Entity
 */
export class InMemoryRepository<T extends Entity> implements Repository<T> {
    private data: Map<T['id'], T> = new Map();
    private nextId = 1;

    /**
     * Find an entity by its ID
     * @param id - Entity ID
     * @returns Promise resolving to the entity or null if not found
     */
    async findById(id: T['id']): Promise<T | null> {
        return this.data.get(id) ?? null;
    }

    /**
     * Get all entities
     * @returns Promise resolving to array of all entities
     */
    async findAll(): Promise<T[]> {
        return Array.from(this.data.values());
    }

    /**
     * Find entities by a specific field value
     * @param field - Field name to search by
     * @param value - Value to match
     * @returns Promise resolving to array of matching entities
     */
    async findBy<K extends keyof T>(field: K, value: T[K]): Promise<T[]> {
        const results: T[] = [];
        for (const entity of this.data.values()) {
            if (entity[field] === value) {
                results.push(entity);
            }
        }
        return results;
    }

    /**
     * Create a new entity
     * @param entity - Entity data without ID
     * @returns Promise resolving to the created entity with ID
     */
    async create(entity: Omit<T, 'id'>): Promise<T> {
        // Generate ID based on the type of the first entity's id or default to number
        let id: T['id'];

        // Check if we have any existing entities to determine ID type
        const firstEntity = this.data.values().next().value;
        if (firstEntity && typeof firstEntity.id === 'string') {
            id = String(this.nextId++) as T['id'];
        } else {
            id = this.nextId++ as T['id'];
        }

        const newEntity = { ...entity, id } as T;
        this.data.set(id, newEntity);
        return newEntity;
    }

    /**
     * Update an existing entity
     * @param id - Entity ID
     * @param updates - Partial entity updates (without ID)
     * @returns Promise resolving to the updated entity or null if not found
     */
    async update(id: T['id'], updates: Partial<Omit<T, 'id'>>): Promise<T | null> {
        const existing = this.data.get(id);
        if (!existing) {
            return null;
        }

        const updated = { ...existing, ...updates } as T;
        this.data.set(id, updated);
        return updated;
    }

    /**
     * Delete an entity by ID
     * @param id - Entity ID
     * @returns Promise resolving to true if deleted, false if not found
     */
    async delete(id: T['id']): Promise<boolean> {
        return this.data.delete(id);
    }

    /**
     * Clear all entities from the repository
     */
    async clear(): Promise<void> {
        this.data.clear();
        this.nextId = 1;
    }

    /**
     * Get the count of entities in the repository
     * @returns Promise resolving to the count
     */
    async count(): Promise<number> {
        return this.data.size;
    }

    /**
     * Check if an entity exists by ID
     * @param id - Entity ID
     * @returns Promise resolving to true if exists, false otherwise
     */
    async exists(id: T['id']): Promise<boolean> {
        return this.data.has(id);
    }
}

File: golang-utilities/generic-utils/repository.go
package genericutils

import (
	"errors"
	"sync"
)

// Entity is the base interface for all entities
type Entity interface {
	GetID() interface{}
}

// Repository defines the interface for CRUD operations
type Repository[T Entity] interface {
	FindByID(id interface{}) (T, error)
	FindAll() []T
	FindBy(predicate func(T) bool) []T
	Create(entity T) (T, error)
	Update(id interface{}, entity T) (T, error)
	Delete(id interface{}) error
	Count() int
	Exists(id interface{}) bool
	Clear()
}

// InMemoryRepository is a thread-safe in-memory implementation of Repository
type InMemoryRepository[T Entity] struct {
	mu     sync.RWMutex
	data   map[interface{}]T
	nextID int
}

// NewInMemoryRepository creates a new in-memory repository
func NewInMemoryRepository[T Entity]() *InMemoryRepository[T] {
	return &InMemoryRepository[T]{
		data:   make(map[interface{}]T),
		nextID: 1,
	}
}

// FindByID finds an entity by its ID (thread-safe)
func (r *InMemoryRepository[T]) FindByID(id interface{}) (T, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	entity, exists := r.data[id]
	if !exists {
		var zero T
		return zero, errors.New("entity not found")
	}
	return entity, nil
}

// FindAll returns all entities (thread-safe)
func (r *InMemoryRepository[T]) FindAll() []T {
	r.mu.RLock()
	defer r.mu.RUnlock()

	entities := make([]T, 0, len(r.data))
	for _, entity := range r.data {
		entities = append(entities, entity)
	}
	return entities
}

// FindBy finds entities matching a predicate (thread-safe)
func (r *InMemoryRepository[T]) FindBy(predicate func(T) bool) []T {
	r.mu.RLock()
	defer r.mu.RUnlock()

	results := make([]T, 0)
	for _, entity := range r.data {
		if predicate(entity) {
			results = append(results, entity)
		}
	}
	return results
}

// Create adds a new entity to the repository (thread-safe)
func (r *InMemoryRepository[T]) Create(entity T) (T, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	id := entity.GetID()
	if id == nil {
		var zero T
		return zero, errors.New("entity ID cannot be nil")
	}

	if _, exists := r.data[id]; exists {
		var zero T
		return zero, errors.New("entity with this ID already exists")
	}

	r.data[id] = entity
	return entity, nil
}

// Update updates an existing entity (thread-safe)
func (r *InMemoryRepository[T]) Update(id interface{}, entity T) (T, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.data[id]; !exists {
		var zero T
		return zero, errors.New("entity not found")
	}

	r.data[id] = entity
	return entity, nil
}

// Delete removes an entity by ID (thread-safe)
func (r *InMemoryRepository[T]) Delete(id interface{}) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.data[id]; !exists {
		return errors.New("entity not found")
	}

	delete(r.data, id)
	return nil
}

// Count returns the number of entities (thread-safe)
func (r *InMemoryRepository[T]) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.data)
}

// Exists checks if an entity exists by ID (thread-safe)
func (r *InMemoryRepository[T]) Exists(id interface{}) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	_, exists := r.data[id]
	return exists
}

// Clear removes all entities (thread-safe)
func (r *InMemoryRepository[T]) Clear() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.data = make(map[interface{}]T)
}

// GetAll is an alias for FindAll
func (r *InMemoryRepository[T]) GetAll() []T {
	return r.FindAll()
}

// Save is a convenience method that creates or updates
func (r *InMemoryRepository[T]) Save(entity T) (T, error) {
	id := entity.GetID()
	if r.Exists(id) {
		return r.Update(id, entity)
	}
	return r.Create(entity)
}
